testng-core/src/main/java/org/testng/TestRunner.java

```diff
@@ -807,11 +808,42 @@ public List<IWorker<ITestNGMethod>> createWorkers(List<ITestNGMethod> methods) {
             .testContext(this)
             .listeners(this.m_classListeners.values())
             .build();
-    return AbstractParallelWorker.newWorker(
-            m_xmlTest.getParallel(), m_xmlTest.getGroupByInstances())
-        .createWorkers(args);
+    List<IWorker<ITestNGMethod>> result =
+        AbstractParallelWorker.newWorker(m_xmlTest.getParallel(), m_xmlTest.getGroupByInstances())
+            .createWorkers(args);
+    long dataDrivenTestCount =
+        result.stream()
+            .flatMap(it -> it.getTasks().stream())
+            .filter(ITestNGMethod::isDataDriven)
+            .count();
+    int threads = getCurrentXmlTest().getThreadCount();
+    XmlSuite.ParallelMode parallelMode = getCurrentXmlTest().getParallel();
+    XmlSuite suite = getSuite().getXmlSuite();
+    if (suite.useGlobalThreadPool()
+        && parallelModes.contains(parallelMode)
+        && dataDrivenTestCount >= threads) {
+      String msg =
+          "[Deadlock condition detected] "
+              + "Cannot run "
+              + dataDrivenTestCount
+              + " data driven tests on just "
+              + threads
+              + " threads when "
+              + "using common thread pool. "
+              + "Please increase the number of threads to at-least "
+              + (dataDrivenTestCount + 1)
+              + ".";
+      throw new TestNGDeadLockException(msg);
+    }
+    return result;
   }
 
+  private static final List<XmlSuite.ParallelMode> parallelModes =
```

Comment:
Could be Set if order is not important

Reply:
Can use EnumSet.allOf(...).filter(::isParallel())

Reply:
I think i could have just used `parallelMode.isParallel()` (didn't realise that this attribute is baked into the enum)